### PR TITLE
Apply soft and hard limits based on the usage of indavidual disk usage rather than aggregate usage

### LINKED ISSUE #2: Improve support for JBOD storage
The storage quota doesn't have currently a very good support for JBOD storage. It sums up the used space across all log directories and evaluates it against the quota. But that does not take into account that one disk might be getting full while other is empty because they have different partitions assigned. Therefore you can still run out of disk space without actually breaching the quota.

### PR DESCRIPTION
**Pull request by SamBarker**
Kafka will take replicas offline if the partition hosting them is out of disk space (see [KIP-112](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD)) so this PR changes the plug-in to apply the soft and hard limits if any disk breaches the limits instead of the aggregate storage. 

In addition this PR extracts a [`QuotaPolicy`](https://github.com/strimzi/kafka-quotas-plugin/blob/e6b4d7c9797d44f21cb75f44c36c65d5ecfe86da/src/main/java/io/strimzi/kafka/quotas/policy/QuotaPolicy.java) out from the `StaticQuotaCallback` to enable different styles of configuring the limits. 

Given this already a breaking change for users with multiple storage volumes (as the way in which the limits are applied changes) it seemed like a good time to try and offer some alternatives configuration options. 

From my perspective it seems clearer to think about the limits in terms of the remaining free space on the volume. However there is not a single obvious term to express that in. There is a case to be made for both min bytes free and min free percentage. Extracting a `QuotaPolicy` allows us to express all the options and to combine them as users want. By being able to compose the policies together it allows to take the conservative option (apply the policy that throttles first) in the face of potentially conflicting configuration. 

To support the min free options new config properties have been added which leads to a bit of [additional logic](https://github.com/strimzi/kafka-quotas-plugin/pull/23/files#diff-70b8d6d689e78856fa72a5e0a459680c1732f191a326eb8d1e26b472fb5cde9fR108-R130) in the  `StaticQuotaConfig`

The majority of the diff is in unit test coverage, the changes to the production code are much simpler than the overall size of the diff implies 😁 

Fixes #2.

### CODE REVIEW

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java [hunk 1]
```diff
@@ -77,13 +80,25 @@ public void run() {
         }
     }
 
+    @Deprecated(forRemoval = true, since = "0.3.0")
```
Comment (SamBarker) [inline src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:84]: I've left this in for API compatibility. 

Happy to delete if that is not needed.

File: src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java [hunk 2]
```diff
@@ -4,16 +4,6 @@
  */
 package io.strimzi.kafka.quotas;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (SamBarker) [inline src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java:7]: Import order was inconsistent between files, to keep my editor happy I've made it consistent. 

Not sure if import order should be covered by checksytle rules?

File: src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java [hunk 3]
```diff
@@ -161,6 +173,229 @@ void staticQuotaMetrics() {
         assertEquals(expectedMbeanName, name.getMBeanName(), "unexpected mbean name");
     }
 
+    @Test
+    void shouldSetQuotaFactorToZeroIfHardLimitBreached() {
+        //Given
+        final Map<String, Long> diskUsage = Map.of("Disk one", 20L);
+        target.configure(Map.of(
```
Comment (SamBarker) [inline src/test/java/io/strimzi/kafka/quotas/StaticQuotaCallbackTest.java:180]: This should maybe done in the `setup` method, I've left it local to the new tests for now.

File: src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java [hunk 4]
```diff
@@ -73,14 +76,12 @@ public Double quotaLimit(ClientQuotaType quotaType, Map<String, String> metricTa
...
         // Don't allow producing messages if we're beyond the storage limit.
         long currentStorageUsage = storageUsed.get();
-        if (ClientQuotaType.PRODUCE.equals(quotaType) && currentStorageUsage > storageQuotaSoft && currentStorageUsage < storageQuotaHard) {
-            double minThrottle = quotaMap.getOrDefault(quotaType, Quota.upperBound(Double.MAX_VALUE)).bound();
-            double limit = minThrottle * (1.0 - (1.0 * (currentStorageUsage - storageQuotaSoft) / (storageQuotaHard - storageQuotaSoft)));
+        if (ClientQuotaType.PRODUCE.equals(quotaType)) {
+            double minThrottle = quotaMap.getOrDefault(quotaType, Quota.upperBound(currentQuotaFactor)).bound();
+            final double produceQuota = minThrottle * currentQuotaFactor;
+            final double limit = Math.max(produceQuota, 1.0D);
             maybeLog(lastLoggedMessageSoftTimeMs, "Throttling producer rate because disk is beyond soft limit. Used: {}. Quota: {}", storageUsed, limit);
```
Comment (SamBarker) [inline src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java:83]: Log message is now slightly miss leading, as the usage could be beyond the hard limit too.

File: src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java [hunk 5]
```diff
@@ -73,14 +76,12 @@ public Double quotaLimit(ClientQuotaType quotaType, Map<String, String> metricTa
...
             return limit;
-        } else if (ClientQuotaType.PRODUCE.equals(quotaType) && currentStorageUsage >= storageQuotaHard) {
-            maybeLog(lastLoggedMessageHardTimeMs, "Limiting producer rate because disk is full. Used: {}. Limit: {}", storageUsed, storageQuotaHard);
```
Comment (SamBarker) [inline src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java:82]: Hard limit application is still logged when calculating the `quotaFactor`

**Review by SamBarker — COMMENTED** [review]
Adding some context for the changes.

File: src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java [hunk 6]
```diff
@@ -180,13 +178,37 @@ public Long value() {
         });
     }
 
-    private void updateUsedStorage(Long newValue) {
-        var oldValue = storageUsed.getAndSet(newValue);
-        if (oldValue != newValue) {
+    //@VisiableForTesting
+    void calculateQuotaFactor(Map<String, Long> usagePerDisk) {
```
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/quotas/StaticQuotaCallback.java:182]: We often do this to make it visible for testing:

```suggestion
    /* test */ void calculateQuotaFactor(Map<String, Long> usagePerDisk) {
```

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java [hunk 7]
```diff
@@ -77,13 +81,25 @@ public void run() {
...
```
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:84]: TBH, I do not think this is any kind of public API. And even if it is, it is still 0.x` version, so I would say unless it is needed for anything, we should just remove it.

**Review by scholzj — COMMENTED** [review]
One of the challenges with supporting multile disks is that it needs a properly configured limit. You indeed cannot handle two 1TB disks as a sum and compare them to a single limit. But you cannot treat the limits as _per-disk_ either since the disks might differ in size.

This is something we looked into in the past. Defining the limit per-sisk seemed a bit cumbersome. So in #10 we instead started working on adding the support for defining the limit as percentage instead. That is something what can easily cover more disks -> 90% soft and 95% hard limit can easily apply to one 1TB and second 4TB disk. The reason this was stopped was that while working on it we realized this does not deal so well with the replication traffic and needs to be redesigned anyway.

If we want to have some better JBOD support now, it is fine, but we need to solve the configurability together with it one way or another.

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java [hunk 8]
```diff
@@ -77,13 +81,25 @@ public void run() {
...
     long checkDiskUsage() {
+        return totalDiskUSage(gatherDiskUsage());
+    }
+
+    long totalDiskUSage(Map<String, Long> diskUsage) {
```
Comment (k-wall) [inline src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:89]: mis-capitalisation in method name.

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java [hunk 9]
```diff
@@ -77,13 +81,25 @@ public void run() {
...
                 .filter(Files::exists)
                 .map(path -> apply(() -> Files.getFileStore(path)))
                 .distinct()
-                .mapToLong(store -> apply(() -> store.getTotalSpace() - store.getUsableSpace()))
-                .sum();
+                .map(store -> {
+                    final Long usedSpace = apply(() -> store.getTotalSpace() - store.getUsableSpace());
+                    return new AbstractMap.SimpleImmutableEntry<>(store.name(), usedSpace != null ? usedSpace : 0);
+                })
+                .collect(Collectors.toMap(AbstractMap.SimpleImmutableEntry::getKey, AbstractMap.SimpleImmutableEntry::getValue));
```
Comment (k-wall) [inline src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:102]: I think you can say the following and achieve the same (untested).
```suggestion
.collect(Collectors.toMap(FileStore::name, StorageChecker::calculateUsedSpace));
```

where
```
    private static long calculateUsedSpace(FileStore store) {
        Long usedSpace = apply(() -> store.getTotalSpace() - store.getUsableSpace());
        return usedSpace != null ? usedSpace : 0;
    }
```

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java [hunk 10]
```diff
@@ -77,13 +81,25 @@ public void run() {
...
+    Map<String, Long> gatherDiskUsage() {
         return logDirs.stream()
                 .filter(Files::exists)
                 .map(path -> apply(() -> Files.getFileStore(path)))
                 .distinct()
-                .mapToLong(store -> apply(() -> store.getTotalSpace() - store.getUsableSpace()))
-                .sum();
+                .map(store -> {
+                    final Long usedSpace = apply(() -> store.getTotalSpace() - store.getUsableSpace());
+                    return new AbstractMap.SimpleImmutableEntry<>(store.name(), usedSpace != null ? usedSpace : 0);
```
Comment (k-wall) [inline src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:100]: Curiosity - how does FileStore.name() relate to the a name kube, if at all?

**Comment by k-wall** [comment]
> One of the challenges with supporting multile disks is that it needs a properly configured limit. You indeed cannot handle two 1TB disks as a sum and compare them to a single limit. But you cannot treat the limits as _per-disk_ either since the disks might differ in size.
> 
> This is something we looked into in the past. Defining the limit per-sisk seemed a bit cumbersome. So in #10 we instead started working on adding the support for defining the limit as percentage instead. That is something what can easily cover more disks -> 90% soft and 95% hard limit can easily apply to one 1TB and second 4TB disk. The reason this was stopped was that while working on it we realized this does not deal so well with the replication traffic and needs to be redesigned anyway.

Percentage based soft and hard limits seem sensible to me.  

To maintain backward compatibility, to bridge from deprecated absolute hard/soft limit, could we not apportion those values across the disks, in proportion  to their size(s)?
> 
> If we want to have some better JBOD support now, it is fine, but we need to solve the configurability together with it one way or another.

**Comment by SamBarker** [comment]
> One of the challenges with supporting multile disks is that it needs a properly configured limit. You indeed cannot handle two 1TB disks as a sum and compare them to a single limit. But you cannot treat the limits as _per-disk_ either since the disks might differ in size.
>
> This is something we looked into in the past. Defining the limit per-sisk seemed a bit cumbersome. So in #10 we instead started working on adding the support for defining the limit as percentage instead. That is something what can easily cover more disks -> 90% soft and 95% hard limit can easily apply to one 1TB and second 4TB disk. The reason this was stopped was that while working on it we realized this does not deal so well with the replication traffic and needs to be redesigned anyway.
>

One thing I keep musing about is: what is the best way to express the limit? Currently it's expressed as bytes *used*. Which as already pointed out doesn't work well with variable disk sizes. Percent full is better but potentially leads to a lot of waisted disk space (I don't remember where that was pointed out). 

When thinking about throttling I've tended to think of applying the throttle when the *free* space is under X we should throttle. I wonder if we can get away with a pair of *new* config properties which express the limits in terms of bytes/percent free. 

The other part of my musing was it makes sense to the throttle producers at X% free, but given the hard limit is a final safety measure I wonder if that is best expressed in bytes.  
E.g. 
- throttle at 5% free
-  and Pause at 10gb free.
  
> If we want to have some better JBOD support now, it is fine, but we need to solve the configurability together with it one way or another.

We have some work happening internally to try and address the replication question, which if successful we would like to convert to a improvement proposal and contribute here. 

I am hopeful that this is change is useful in its own right while making that work easier.

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java (hunk 9 above)
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:102]: Ohh nice.

File: src/main/java/io/strimzi/kafka/quotas/StorageChecker.java (hunk 10 above)
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/quotas/StorageChecker.java:100]: I need to fire it up in K8s to check. The FileStore documentation is quite vague on what it really represents.

**Comment by scholzj** [comment]
> One thing I keep musing about is: what is the best way to express the limit? Currently it's expressed as bytes _used_. Which as already pointed out doesn't work well with variable disk sizes. Percent full is better but potentially leads to a lot of waisted disk space (I don't remember where that was pointed out).
> 
> When thinking about throttling I've tended to think of applying the throttle when the _free_ space is under X we should throttle. I wonder if we can get away with a pair of _new_ config properties which express the limits in terms of bytes/percent free.
> 
> The other part of my musing was it makes sense to the throttle producers at X% free, but given the hard limit is a final safety measure I wonder if that is best expressed in bytes. E.g.
> 
> * throttle at 5% free
> * and Pause at 10gb free.

Yes, percentage used (or percentage free) are not optimal. But there will be for sure many cases where _bytes free_ would not be optimal either. TBH, I do not think they are any better or worse compared to each other. So I would be fine with either of them. So if you prefer _bytes free_ we can go for it. I would probably not mix percentage and bytes free for the soft and hard limits. It would be less fool proof and harder to validate that soft limit is before the hard limit.

> We have some work happening internally to try and address the replication question, which if successful we would like to convert to a improvement proposal and contribute here.
> 
> I am hopeful that this is change is useful in its own right while making that work easier.

I think there is a big risk of a wasted effort. That is why we stopped the work at #10. In general, if you want to invest your effort into it now, that is fine. but I think we need the proper configurability for JBOD together with this.

**Comment by tombentley** [comment]
I agree that limits based on used space was, in hindsight, a mistake. I think we could support both byte and percentage _free_ space (and if both are specified we use whichever is biggest). We'd need to support the old way too, for backwards compatibilty, but we should say that that is deprecated and log a warning to that effect. I would also allow a global config for this, plus being able to override on a per-logdir basis.

**Comment by SamBarker** [comment]
> I agree that limits based on used space was, in hindsight, a mistake. I think we could support both byte and percentage _free_ space (and if both are specified we use whichever is biggest). We'd need to support the old way too, for backwards compatibilty, but we should say that that is deprecated and log a warning to that effect. I would also allow a global config for this, plus being able to override on a per-logdir basis.

I've implemented min free support, but haven't done per volume configuration as I don't see what use case that would support. 

Because I haven't built per volume configuration I haven't added validation to select the more conservative of free bytes or free Percentage (when both configured), however In the process of writing this comment I may have just worked out a simple enough way to support that and simplify the quota policy configuration method....

File: src/main/java/io/strimzi/kafka/quotas/VolumeDetails.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2021, Red Hat Inc.
```
Comment (scholzj) [inline src/main/java/io/strimzi/kafka/quotas/VolumeDetails.java:2]: I don't think you should be including any Red Hat copyright headers in new files. You should use something like this:

```
/*
 * Copyright Strimzi authors.
 * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
 */
```

**Review by scholzj — COMMENTED** [review]
I didn't went through the code in detail. But perhaps you can a bit more explain what exactly you did? We will also need to update the README.md file (maybe it is something to do at the end when it is clear how it ends up, but we should not forget it). It would be also good to have some JavaDocs though the code. E.g. explaning the policies etc. I also wonder whenr more discussion is needed to which policies should be supported rather than just supporting everything.

**Comment by scholzj** [comment]
> I've implemented min free support, but haven't done per volume configuration as I don't see what use case that would support. I'm keen to understand when a min free approach, as either bytes or percentage, isn't enough and we need per volume granularity.

For me, the limit where I want to start slowing down the data or stop them completely is based on _time until collapse_. nd that is basically a function of throughput. With high throughput, I would set the limit more carefully, but with low throughput I would be less careful. With 100MB/s, 100GB of free space can fill in ~15 minutes and might not be enough. With 100KB/s it will be ~11 days which is plenty of time.

At the same time, the throughput does not necessarily determine the  disk size -> you can have use-cases such as log or metric collection which is hight throughput but often relatively short retention. And you can have for example some EventSourcing patterns which are collecting data for years, but overall throughput might not be huge.

While it can be discussed whether it is smart and good move, sometimes, use-cases like these are combined in one cluster / broker. And you can see for example people using combinations of different disk types (e.g. slower and cheaper `st1` EBS storage with faster and more expensive `gp` or `io` volumes.)

This might drive the requirement to have the different limits for different disks (or in general even different brokers). That said, for me personally, it would not be a must-have for this PR.

File: src/main/java/io/strimzi/kafka/quotas/VolumeDetails.java (hunk 11 above)
Comment (SamBarker) [reply src/main/java/io/strimzi/kafka/quotas/VolumeDetails.java:2]: Thanks, I did wonder about that but was copying the existing files. 

I'll update my changes to use strimizi authors.

**Comment by scholzj** [comment]
PS: While I appreciate the effort put into this PR, I think it is a bit concerning at the same time. The reason why we didn't proceed with #10 was the lack of plan how to deal with the bigger issue which was the lack of handling for the replication data. We still didn't moved anywhere with it. Yet, we seem to be coming with multiple new ways how to configure the plugin and deprecating the old options. There is a risk that whatever the solution is to the replication issue might significantly affect the plugin's design. Ideally, we want to avoid deprecating old options and adding new options all the time and rather find the proper solution for the key issues first.

**Comment by SamBarker** [comment]
Yeah I agree this PR has moved away from its original intent, which was simply focused on improving support for multiple disks. The scope has crept to include defining min free policies, the associated configuration parameters and the added complexity of applying two different styles of policy.

I had never considered using different classes of storage for different log directories and that does present an interesting use case for having per volume configuration. I would lean towards addressing that in a separate PR. 

I'll update the PR description to outline the changes made, and will update the readme once we are settled on the configuration changes.

**Comment by SamBarker** [comment]
I've opened a https://github.com/strimzi/proposals/pull/51 to address the replication issues. I think the PR supports that effort, but I think its strayed enough from its original intent to be blocked on agreement to that proposal.

**Comment by SamBarker** [comment]
Following the rational for closing https://github.com/strimzi/proposals/pull/51 this PR no longer makes sense to keep open. 

I will return to the quota plug-in changes once KIP-827 is a practical option.

**Pull request closed without merging**